backend/common/src/main/java/ai/verta/modeldb/common/reconcilers/Reconciler.java

```diff
@@ -15,23 +16,28 @@
 import java.util.concurrent.locks.Condition;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
 
 public abstract class Reconciler<T> {
   final Logger logger;
   protected final HashSet<T> elements = new HashSet<>();
   protected final LinkedList<T> order = new LinkedList<>();
   final Lock lock = new ReentrantLock();
   final Condition notEmpty = lock.newCondition();
+  // To prevent OptimisticLockException
+  private final Set<T> processingIdSet = ConcurrentHashMap.newKeySet();
```

Comment:
As per the the conversation in slack, this is not enough to make things thread safe.

Reply:
`ConcurrentHashMap.newKeySet();` will create the thread safe `Set`  so it is enough as per my thought.

Reply:
It is not. Please refer to the slack thread where I described the 2 race conditions. This fixes one (as I described in the thread) but not the other